rubullet/src/client.rs

```diff
@@ -0,0 +1,3254 @@
+    /// * [get_camera_image](`Self::get_camera_image`)
+    pub fn compute_projection_matrix(
+        left: f32,
+        right: f32,
+        bottom: f32,
+        top: f32,
+        near_val: f32,
+        far_val: f32,
+    ) -> [f32; 16] {
+        let mut projection_matrix = [0_f32; 16];
+        unsafe {
+            ffi::b3ComputeProjectionMatrix(
+                left,
+                right,
+                bottom,
+                top,
+                near_val,
+                far_val,
+                projection_matrix.as_mut_ptr(),
+            );
+            projection_matrix
+        }
+    }
+    /// computes the projection matrix which can be used together with the view matrix to generate
+    /// camera images within the simulation
+    ///
+    /// # Arguments
+    /// * `fov` - left screen (canvas) coordinate
+    /// * `aspect` - right screen (canvas) coordinate
+    /// * `near_val` - near plane distance
+    /// * `far_val` - far plane distance
+    ///
+    /// # See also
+    /// * [compute_view_matrix](`Self::compute_view_matrix`)
+    /// * [compute_view_matrix_from_yaw_pitch_roll](`Self::compute_view_matrix_from_yaw_pitch_roll`)
+    /// * [compute_projection_matrix](`Self::compute_projection_matrix`)
+    /// * [get_camera_image](`Self::get_camera_image`)
+    pub fn compute_projection_matrix_fov(
+        fov: f32,
+        aspect: f32,
+        near_val: f32,
+        far_val: f32,
+    ) -> [f32; 16] {
+        let mut projection_matrix = [0_f32; 16];
+        unsafe {
+            ffi::b3ComputeProjectionMatrixFOV(
+                fov,
+                aspect,
+                near_val,
+                far_val,
+                projection_matrix.as_mut_ptr(),
+            );
+            projection_matrix
+        }
+    }
+    /// returns an RGBA, depth and segmentation images.
+    ///
+    /// # Note
+    /// Depth and segmentation images are currently not really images as the image crate does not
+    /// properly support these types yet.
+    ///
+    /// # Arguments
+    /// * `width` - eye position in Cartesian world coordinates
+    /// * `height` - position of the target (focus) point, in Cartesian world coordinates
+    /// * `view_matrix` -  view matrix, see [compute_view_matrix](`Self::compute_view_matrix`)
+    /// * `projection_matrix` - projection matrix, see [compute_projection_matrix](`Self::compute_projection_matrix`)
+    /// # See also
+    /// * [compute_view_matrix](`Self::compute_view_matrix`)
+    /// * [compute_view_matrix_from_yaw_pitch_roll](`Self::compute_view_matrix_from_yaw_pitch_roll`)
+    /// * [compute_projection_matrix](`Self::compute_projection_matrix`)
+    /// * [compute_projection_matrix_fov](`Self::compute_projection_matrix_fov`)
+    /// * panda_camera_demo.rs for an example
+    pub fn get_camera_image(
+        &mut self,
+        width: i32,
+        height: i32,
+        view_matrix: &[f32; 16],
+        projection_matrix: &[f32; 16],
```

Comment:
So it seems like that view matricies are homogeneous matrices but projection matrices are not. A case for not using the isometry is that the conversion from Matrix4 to Isometry3 an back could slightly alter the matrix as it gets normalized. Maybe bullet doesnt like that. Affine3 should be able to do this without a problem. However, it seems like it is not that well documented and I have not seen somebody else using it. I would go for Matrix4.

Reply:
View matrices are isometries in all but the weirdest situations, so I would expect most people to be storing them as isometries. However, internally we will be calling `Isometry3::to_homogeneous` so taking a 4x4 matrix wouldn't change much and would avoid conversions for people who have decided to use raw matrices.

Projections are homogeneous matrices. However, I did type the previous comment wrong, implying that the projection matrix should be `Isometry3` when that isn't the case. Additionally, `Affine3` is also not the correct type - assuming "normal" usage, it would be a [`Projective3`][1] ([hierarchy][2]). The use of the projective type would be self-documenting that the projection argument is a projective transformation while allowing both typical transformations (orthographic and perspective) and would still allow arbitrary transformations through abusing `Projective3::from_matrix_unchecked`. 

All of that being said, you are correct that the `Transform<...>` types aren't commonly used or particularly well documented.

[1]: https://docs.rs/nalgebra/0.24.1/nalgebra/geometry/type.Projective3.html
[2]: https://www.nalgebra.org/docs/user_guide/points_and_transformations#transformations

Reply:
I am now using Matrix4 for everything